src/mokernels/moinput.jl

```diff
@@ -18,40 +47,83 @@ julia> MOInput(x, 2)
 The first `length(x)` elements represent the inputs for the first output, the second
 `length(x)` elements represent the inputs for the second output, etc.
-
-See [Inputs for Multiple Outputs](@ref) in the docs for more info.
 """
-struct MOInput{T<:AbstractVector} <: AbstractVector{Tuple{Any,Int}}
+struct MOInputIsotopicByOutputs{S,T<:AbstractVector{S}} <: AbstractVector{Tuple{S,Int}}
     x::T
-    out_dim::Integer
+    out_dim::Int
 end
 
-Base.length(inp::MOInput) = inp.out_dim * length(inp.x)
-
-Base.size(inp::MOInput, d) = d::Integer == 1 ? inp.out_dim * size(inp.x, 1) : 1
-Base.size(inp::MOInput) = (inp.out_dim * size(inp.x, 1),)
-
-Base.lastindex(inp::MOInput) = length(inp)
-Base.firstindex(inp::MOInput) = 1
-
-function Base.getindex(inp::MOInput, ind::Integer)
-    if ind > 0
-        out_dim = ind ÷ length(inp.x) + 1
-        ind = ind % length(inp.x)
-        if ind == 0
-            ind = length(inp.x)
-            out_dim -= 1
-        end
-        return (inp.x[ind], out_dim::Int)
-    else
-        throw(BoundsError(string("Trying to access at ", ind)))
-    end
+const IsotopicMOInputs = Union{MOInputIsotopicByFeatures,MOInputIsotopicByOutputs}
+
+function Base.getindex(inp::MOInputIsotopicByOutputs, ind::Integer)
+    @boundscheck checkbounds(inp, ind)
+    output_index, feature_index = fldmod1(ind, length(inp.x))
+    feature = @inbounds inp.x[feature_index]
+    return feature, output_index
 end
 
-Base.iterate(inp::MOInput) = (inp[1], 1)
-function Base.iterate(inp::MOInput, state)
-    return (state < length(inp)) ? (inp[state + 1], state + 1) : nothing
+function Base.getindex(inp::MOInputIsotopicByFeatures, ind::Integer)
+    @boundscheck checkbounds(inp, ind)
+    feature_index, output_index = fldmod1(ind, inp.out_dim)
+    feature = @inbounds inp.x[feature_index]
+    return feature, output_index
 end
+
+Base.size(inp::IsotopicMOInputs) = (inp.out_dim * length(inp.x),)
+
+"""
+    MOInput(x::AbstractVector, out_dim::Integer)
+A data type to accomodate modelling multi-dimensional output data.
+`MOInput(x, out_dim)` has length `length(x) * out_dim`.
+
+```jldoctest
+julia> x = [1, 2, 3];
+
+julia> MOInput(x, 2)
+6-element KernelFunctions.MOInputIsotopicByOutputs{Int64, Vector{Int64}}:
+ (1, 1)
+ (2, 1)
+ (3, 1)
+ (1, 2)
+ (2, 2)
+ (3, 2)
+```
+As shown above, an `MOInput` represents a vector of tuples.
+The first `length(x)` elements represent the inputs for the first output, the second
+`length(x)` elements represent the inputs for the second output, etc.
+See [Inputs for Multiple Outputs](@ref) in the docs for more info.
+"""
+const MOInput = MOInputIsotopicByOutputs
+
+"""
+    isotopic_by_outputs(x::AbstractVector, out_dim::Integer)
+
+Helper function to construct [`IsotopicByOutputs`](@ref).
+
+```jldoctest isotopic_by_outputs
+julia> x = [1, 2, 3];
+
+julia> KernelFunctions.isotopic_by_outputs(x, 2) == KernelFunctions.MOInputIsotopicByOutputs(x, 2)
+true
+```
+"""
+const isotopic_by_outputs = MOInputIsotopicByOutputs
```

Comment:
it's too early, I belatedly realized that this might also require a thin wrapper function instead of the const alias:
```suggestion
isotopic_by_outputs(features, output_dim) = MOInputIsotopicByOutputs(features, output_dim)
```